src/main/java/prover/Prover.java

```diff
@@ -0,0 +1,195 @@
+import java.io.IOException;
+import java.util.ArrayList;
+import org.apache.commons.cli.*;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.api.java.function.Function2;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.storage.StorageLevel;
+import scala.Tuple2;
+import zk_proof_systems.zkSNARK.grothBGM17.DistributedProver;
+import zk_proof_systems.zkSNARK.grothBGM17.ZKSnarkObjectReader;
+import zk_proof_systems.zkSNARK.grothBGM17.ZKSnarkObjectWriter;
+
+public class Prover {
+  public static void main(String[] args) throws IOException {
+    System.out.println("Distributed Prover (PoC implementation)");
+
+    var options = new Options();
+    options.addOption(new Option("h", "help", false, "Display this message"));
+    options.addOption(new Option("l", "local", false, "Run on local simulated cluster"));
+    options.addOption(new Option("t", "test", false, "Run trivial test to verify setup"));
+    options.addOption(new Option("p", "primary-size", false, "Size of primary input (1)"));
+    options.addOption(new Option("o", "output", true, "Output file (proof.bin)"));
+
+    try {
+      var parser = new BasicParser();
+      var cmdLine = parser.parse(options, args);
+      var trailing = cmdLine.getArgs();
+      System.out.println("trailing: " + String.join(", ", trailing));
+
+      if (cmdLine.hasOption("help")) {
+        print_usage(options);
+        return;
+      }
+
+      if (cmdLine.hasOption("test")) {
+        runTest(cmdLine.hasOption("local"));
+        return;
+      }
+
+      final int primaryInputSize = Integer.parseInt(cmdLine.getOptionValue("primary-size", "1"));
+      final String outputFile = cmdLine.getOptionValue("output", "proof.bin");
+
+      // Extract command lin arguments and call run.
+      if (trailing.length != 2) {
+        System.err.println("error: invalid number of arguments\n");
+        print_usage(options);
+        System.exit(1);
+      }
+
+      runBN254a(primaryInputSize, trailing[0], trailing[1], outputFile, cmdLine.hasOption("local"));
+    } catch (ParseException e) {
+      System.err.println("error: " + e.getMessage());
+    }
+  }
+
+  static void print_usage(Options options) {
+    new HelpFormatter().printHelp("prover <PROVING-KEY-FILE> <ASSIGNMENT-FILE>", options);
+  }
+
+  static JavaSparkContext createSparkContext(boolean local) {
+    final var sessionBuilder = SparkSession.builder().appName("prover");
+
+    if (local) {
+      sessionBuilder.master("local");
+    }
+
+    final SparkSession spark = sessionBuilder.getOrCreate();
+
+    spark.sparkContext().conf().set("spark.files.overwrite", "true");
+    // spark.sparkContext().conf().set("spark.serializer",
+    // "org.apache.spark.serializer.KryoSerializer");
+    // spark.sparkContext().conf().registerKryoClasses(SparkUtils.zksparkClasses());
+    return new JavaSparkContext(spark.sparkContext());
+  }
+
+  static void runTest(boolean local) throws IOException {
```

Comment:
I am not sure I get the purpose of this function. I can see that it is ran when the flag `--test` is set. What setup do you verify here?

Reply:
This is a simple test that the cluster is working and communicating correctly, which can be run standalone without any test data etc.

Reply:
Got it, thanks